Implement DataSource to raw datapackage method

zschira opened the pull request:
This pull request enables integration of the new `DataSource` metadata models to generate datapackage metadata used by [pudl-zenodo-storage](https://github.com/catalyst-cooperative/pudl-zenodo-storage) ([relevant pull request](https://github.com/catalyst-cooperative/pudl-zenodo-storage)).

zschira commented on src/pudl/metadata/classes.py:
```diff
@@ -933,9 +933,24 @@ def raw_datapackage_title(self) -> str:
+        """Construct datapackage sources for raw data source."""
+        return [{"title": self.title, "name": self.path}]
+
+    def to_raw_datapackage_dict(self) -> Dict:
```
I changed this to return a dict because it integrates a bit cleaner into the `pudl-zenodo-storage` code, but we could return JSON directly if that'd be preferred.

zaneselvans replied:
That makes sense.

One thing that has been a little bit frustrating with the Zenodo storage setup is that there's no way to preview the `datapackage.json` file that it's going to output without actually going to the trouble of uploading all of the data to Zenodo. If it was easy to have a "dry run" or "no op" version of the command that output that JSON file or printed it out that could be helpful.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -933,9 +933,24 @@ def raw_datapackage_title(self) -> str:
-    def to_raw_datapackage_json(self) -> pydantic.Json:
+    def raw_datapackage_sources(self) -> List[Dict[str, str]]:
+        """Construct datapackage sources for raw data source."""
+        return [{"title": self.title, "name": self.path}]
```
In the [datapackage specification](https://specs.frictionlessdata.io/data-package/#sources), it looks like in the individual `source` elements what's called `name` here should probably be called `path`. Does that seem right?

zschira replied:
Yeah I think that's correct. Fixed now

zaneselvans commented:
I added a unit test to check that all of the stored data sources can actually be instantiated (and added descriptions to the ones that didn't have them, so they'd validate)

The information that's stored in the `DataSource` class isn't enough to create a valid data package descriptor. Where should that logic live? It seems like there's at least two clear optiosn:

* There could be a data model in the Zenodo repo that instantiates the relevant `DataSource` and can access its fields and construct the descriptor using other data it also contains.
* We could pass that other information into `DataSource.to_raw_datapackage_descriptor()` such that the `DataSource` has all the information it needs to return a valid descriptor. This would include a list of resource descriptors and probably other information too.

In the end I would expect something like this to pass and we should probably put it in the `pudl-zenodo-storage` repo tests for the data sources we expect to be able to archive, and add a dummy version to the PUDL unit tests:

```py
from datapackage import Package
from pudl.metadata.classes import DataSource
for source_id in pudl.metadata.sources.SOURCES:
    data_source = DataSource.from_id(source_id)
    # Seems like maybe we need to pass in additional information so that
    # this method can construct the complete datapackage descriptor.
    data_pkg = Package(descriptor=data_source.to_raw_datapackage_dict())
    if not data_pkg.valid:
        # raise / display information in data_pkg.errors
        raise
```

zaneselvans commented:
But there's kind of already something that has the job of building the datapackage descriptors on the `pudl-zenodo-storage` side of things, all the `datapackager` functions that live in the `frictionless.core` module. They look very amenable to being converted into Pydantic models, and that would let us keep all the knowledge about the internal datapackage construction over on the zenodo side of things.

zschira commented:
Moved logic regarding generating raw datapackage descriptors to `pudl-zenodo-storage`. Also added `keywords` to `censusdp1tract` and `ferc714` sources to enable generation of valid datapackage descriptors.

zaneselvans reviewed (approved):
Looks good.
